File: javapatterns/src/main/java/org/iut/nantes/decorator/nombres/Decorator.java
package org.iut.nantes.decorator.nombres;

abstract class Decorator extends NombreAbstrait
{
  protected NombreAbstrait noyau;

  public Decorator( NombreAbstrait nb )
  {
    noyau = nb;
  }

  public int getEntier()
  {
    return noyau.getEntier();
  }
}

File: kotlinpatterns/src/main/kotlin/org/iut/nantes/decorator/concession/Decorator.kt
package org.iut.nantes.decorator.concession

abstract class Decorator(val serie: Serie) : Serie() {
    abstract var bonus: Float

    override var cout: Float
        get() = serie.cout + bonus
        set(value) {}

    override var carburant: Carburant = serie.carburant
}

class PeintureMetallisee(serie: Serie) : Decorator(serie) {
    override var bonus: Float
        get() = 1500f
        set(value) {}

}

class SiegeEnCuir(serie: Serie) : Decorator(serie) {
    override var bonus: Float
        get() = 2000f
        set(value) {}
}

class GPS(serie: Serie): Decorator(serie) {
    override var bonus: Float
        get() = 5200f
        set(value) {}
}

